### ISSUE TITLE
Feature Request: Add data binding support for default interface properties (C# 8)

### ISSUE THREAD

**Issue by zlatanov** [dotnet/wpf#2959]
C# 8 added support for default interface method and properties. Now we can write the following:
``` c#
public interface ITimeFeature
{
    public DateTime CurrentTime => DateTime.Now;
}
class Item : ITimeFeature {}
```

However if we try to data bind to the **Item** object with **Path=CurrentTime**, it will fail to locate the property because it is inherited from the interface and not explicitly defined.

We can easily make this work now by creating a **CustomTypeDescriptor** which includes the defaulted properties. Can you make this built in, so we don't have to resort to writing descriptors?

**Comment by weltkante** [comment]
> Path=CurrentTime

That looks at the type of the object so if the property only exists on an interface it is intentional that this doesn't work, otherwise you'd get _a lot_ of ambiguities. Does this work if you use the qualified syntax `Path=(local:ITimeFeature.CurrentTime)` to tell the engine to look at the interface instead of the class?

**Comment by zlatanov** [comment]
@weltkante you need to access it through the interface. The property really doesn't exist in the type. 

And yes, fully qualifying the property works. But that is too much of a hassle and in some situations is not good enough - group by descriptors, sort member paths, etc.

**Comment by miloush** [comment]
@zlatanov if it's an explicit interface member, then the requirement to fully qualify it seems justified. How would you expect multiple inheritance conflicts to be resolved otherwise?

**Comment by zlatanov** [comment]
@miloush When the implementation of the interface is explicit (in regards to that property) I expect to receive data binding error ( property not found ). If more than one implemented interfaces provides default property with the same name, then I again would expect a data binding error ( conflict ).

Conflicts can already happen without interfaces. Imagine if we have the following:

```c#
abstract class Base
{
    public String Id { get; }
}
class X : Base
{
    public new Guid Id { get; }
}
```

Now class X has two properties with the name Id. WPF must and can deal with those in a well defined manner.

**Comment by weltkante** [comment]
Note that this example is not a conflict and has defined rules. Implementing two interfaces with the same property on the same object does _not_ have defined rules, you'd have to make something up to implement this issue. (Not impossible, there's just no precedence in C# to resolve this kind of conflict.)

**Comment by zlatanov** [comment]
@weltkante I have to disagree here. It is a conflict, if you call typeof( X ).GetProperty( "Id" ) it will throw exception because more than 1 property matches the name. It's just that WPF knows how to handle it.

I can however give you a more extreme example. Suppose we have this:
```c#
using System;
using System.Reflection;
using System.Reflection.Emit;
using System.Windows;

namespace WpfApp4
{
    public partial class MainWindow : Window
    {
        public MainWindow()
        {
            var assemblyName = new AssemblyName("dynamic");
            var assembly = AssemblyBuilder.DefineDynamicAssembly(assemblyName, AssemblyBuilderAccess.Run);
            var module = assembly.DefineDynamicModule("dynamic.dll");

            var type = module.DefineType("X", TypeAttributes.Public | TypeAttributes.Sealed);
            var property1 = type.DefineProperty("Id", PropertyAttributes.None, returnType: typeof(string), parameterTypes: null);
            property1.SetGetMethod(DefineGetMethod(type, typeof(string), x =>
            {
                x.Emit(OpCodes.Ldstr, "String");
                x.Emit(OpCodes.Ret);
            }));

            var property2 = type.DefineProperty("Id", PropertyAttributes.None, returnType: typeof(int), parameterTypes: null);
            property2.SetGetMethod(DefineGetMethod(type, typeof(int), x =>
            {
                x.Emit(OpCodes.Ldc_I4, 42);
                x.Emit(OpCodes.Ret);
            }));

            var x = type.CreateType();
            var instance = Activator.CreateInstance(type);

            DataContext = instance;

            InitializeComponent();
        }


        private static MethodBuilder DefineGetMethod(TypeBuilder type, Type propertyType, Action<ILGenerator> generator)
        {
            var method = type.DefineMethod(name: "_getId",
                                           attributes: MethodAttributes.Public | MethodAttributes.HideBySig,
                                           returnType: propertyType,
                                           parameterTypes: null);
            generator(method.GetILGenerator());

            return method;
        }
    }
}
```

Here I have two properties with the same name in the same type. One is `string` and the other is `Guid`. The CLR runtime has no problem with this. Now what do you suppose happens when we try to data bind the "Id" property, what would we expect to see?

I personally would expect to see a databinding error saying that there is ambiguity here and more than one property is found. However the databinding succeeds and we see the text property because it's defined first. If I define the int property we would see it instead.

My point here is that these cases already exist and there is code that tries to reconcile them (in some cases incorrectly).

**Comment by weltkante** [comment]
>  I have to disagree here. It is a conflict, if you call typeof( X ).GetProperty( "Id" ) it will throw exception because more than 1 property matches the name.

I said it has defined rules in C# and data binding, the conflict you quote comes from reflection. What I was saying is your example was bad because there is precedence how to resolve it. There is no precedence how to resolve conflicts in multiple interfaces on the same class, so whatever you make up for implementing this issue will be **WPF only** and not something agreed on across UI frameworks.

The second example you gave is not valid C# (the language forbids these conflicts). Sure, this conflict exists, but its not relevant for what I was saying

> My point here is that these cases already exist and there is code that tries to reconcile them

I doubt that, picking the first from a list instead of throwing does not qualify as code to reconcile the choices.

Alltogether you don need to convince me of anything, if the WPF team wants to integrate support for default interface properties into the databinding I'm not going to stand in their way

**Comment by miloush** [comment]
There is `ObjectDataProvider` however, so that should be consistent in terms of which methods are visible on types.

The C# language team has made an explicit decision that default properties will not be visible on the implementing types, and this seems like an attempt to circumvent that decision.

**Comment by miloush** [comment]
In fact there is already an existing strategy to resolve the conflict, it's the qualified syntax as mentioned above, consistent with how the language requires it to be resolved.

You can also employ a type descriptor/dynamic to make the binding work the way you suggest and again, that sounds like an justified mechanism to me for introducing such change in behavior.

Last but not least, if in the future developers will get a choice to have default properties to be implemented either explicitly or implicitly, the behavior in this feature request would be a broken one, choosing the wrong property to bind to.

**Comment by zlatanov** [comment]
@weltkante I don't think C# has anything to do with this. I could be using VB if I wanted.

See here: https://github.com/dotnet/wpf/blob/ae1790531c3b993b56eba8b1f0dd395a3ed7de75/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/PropertyPath.cs#L661-L672

This is where the magic happens and it is indeed an ambiguity and WPF tries to resolve it. The comment in the catch block is partly true, because there are other ways this can happen. Nothing specific to C# in there.

@miloush I can see your point. However explicit or implicit implementation of the interface is not breaking the databinding, the CLR runtime knows if the type implements the property and uses its body and not the default one.

**Comment by miloush** [comment]
In the case of a fallback lookup, adding a proper property (or allowing implicit implementation in the future) to the type would be a breaking change to WPF applications depending on the behavior picking it up instead of the fallback.

Performance impact should also be considered, as the binding would need to go through all implemented interfaces to check if they define a property. Sometimes that's a couple of interfaces to go through (comparable, equatable, cloneable, property notifications, many more for collections).

**Comment by weltkante** [comment]
@zlatanov thanks, wasn't aware of this, your previous response makes more sense now

**Comment by miloush** [comment]
> // We'll resolve this by returning the most specific property. 

That is a valid approach when there is a notion of more and less specific. In case of multiple inheritance, there is no clear winner. A binding error sounds more appropriate to me here. This also seems to be the decision taken by C# team for [diamond inheritance](https://github.com/dotnet/csharplang/blob/master/meetings/2018/LDM-2018-10-17.md#issues-in-default-interface-methods):

> We've come down to two options: pick an implementation somewhat arbitrarily, or produce a runtime exception. [...] We're going to throw a runtime exception (we want a new exception type) when there is a runtime diamond and there is no unique most derived implementation. 

If the WPF team decided they want to fallback to explicit interface members everywhere whenever there is a binding error, I don't have strictly blocking arguments either, but I think the performance of existing apps can only decrease, that the benefits are simply not worth the work and that it goes against the idea/purpose of explicit implementations. First time you would want to improve performance of your app, the advice would be to use qualified syntax instead of this go-through-interfaces fallback mechanism.

**Comment by zlatanov** [comment]
@miloush True. However I don't think the runtime resolve of a property path is where the performance is worst, because it is done once per type and can be very easily cached, because it cannot change during the execution of the program.

There are so many low hanging fruits for optimizing WPF bindings, it's crazy. I really hope WPF team  opens up more and really start accepting pull requests.
